[gui] fix media source workarounds

tamland opened the pull request:
This basically a workaround to allows using real VFS path as roots in media windows instead of the current workaround of using media sources to emulate it. This is the cause for a lot of performance issues gui quirks, primarily seen in the addon browser.

By using the VFS it will behave as a regular directory without special 'sources' root node.

tamland committed:
[gui] allow using vfs as root in media windows

tamland committed:
[gui] remove 'sources' workarounds addons/event window

tamland commented:
@xhaggi Can you review please?

Razzeee commented on xbmc/windows/GUIMediaWindow.cpp:
```diff
@@ -727,30 +713,36 @@ bool CGUIMediaWindow::GetDirectory(const std::string &strDirectory, CFileItemLis
 // \brief Set window to a specific directory
 // \param strDirectory The directory to be displayed in list/thumb control
```
This would also need to be ```_arg``` I guess? And the second param is missing, but that's not your doing :/

Razzeee commented on xbmc/windows/GUIMediaWindow.cpp:
```diff
@@ -727,30 +713,36 @@ bool CGUIMediaWindow::GetDirectory(const std::string &strDirectory, CFileItemLis
 // \param strDirectory The directory to be displayed in list/thumb control
 // This function calls OnPrepareFileItems()
-bool CGUIMediaWindow::Update(const std::string &strDirectory, bool updateFilterPath /* = true */)
+bool CGUIMediaWindow::Update(const std::string &_arg, bool updateFilterPath /* = true */)
```
Changed parameter name should probably also be changed to match in the header file?

Montellese replied:
Since when do we use variable names starting with an underscore?

tamland replied:
Should we not? I tend to use it for throwaway args to make it clear it should not be used directly, rather the 'fixed' `path` below should. I guess not clear..

xhaggi replied:
-1 for using underscores to mark them as "throwaway" args and using `arg` as parameter name in this case.

We have tons of methods where we use const parameters only used to check something and process with a copy of that parameter. The same you do here and BTW you break the readability of the code because who knows what `_arg` is. To understand the code now, you need to go to the method docs and check the comment where you find "\param _arg The current directory path" .. that's really annoying.

tamland replied:
Then I'll rename it strDirectoryOrEmpty or something unless you have a better suggestion. If this parameter was "The current directory path" I'd agree with you, but it isn't. This argument has at least 3 different meanings here: "?" some magic value for doing nothing, empty string which means get the root, or the actual path.

xhaggi replied:
IMO we should leave it as it is, because each sub class use the same parameter name. Adjust the method doc and make it clear there. We could think about a better naming for it and do it the right way.

Montellese commented on xbmc/windows/GUIMediaWindow.cpp:
```diff
@@ -677,22 +679,6 @@ bool CGUIMediaWindow::GetDirectory(const std::string &strDirectory, CFileItemLis
-  }
-
-  // update the view state's reference to the current items
-  m_guiState.reset(CGUIViewState::GetViewState(GetID(), items));
```
Are you 100% sure that this code isn't needed anymore? IIRC it is/was also needed for the filtering logic which changes the retrieved path. But it's been a while since I looked into it.

xhaggi replied:
@tamland what's the reason why you moved the whole block to `Update()`

tamland replied:
@xhaggi It breaks views state loading for classes overriding it. What you're looking at is a separate commit btw. This should always be done for all directories, it's not and implementation detail of GetDirectory so I moved it Update.

xhaggi commented on xbmc/windows/GUIMediaWindow.cpp:
```diff
@@ -213,11 +213,13 @@ bool CGUIMediaWindow::OnAction(const CAction &action)
-  if (actionID == ACTION_NAV_BACK && !m_vecItems->IsVirtualDirectoryRoot() &&
+  if (actionID == ACTION_NAV_BACK &&
+      //!m_vecItems->IsVirtualDirectoryRoot() &&
+      //!URIUtils::PathEquals(m_vecItems->GetPath(), GetRootPath()) &&
```
leftovers?

tamland committed:
[addons] move 'downloading' directory to vfs

fixes view state not loading correctly as CGUIMediaWindow::GetDirectory is not called

tamland commented:
Updated! I dropped the 'view state' commit as it still didn't entirely fix it. I must be missing something because even if I moved the rest out of GetDirectory it still wouldn't update the view correctly when overriding... I've moved the directories to the vfs layer instead (probably a better solution anyway) such that CGUIMediaWindow::GetDirectory is always called.

tamland commented:
jenkins build this please

tamland commented:
I'll go ahead with this as I've tested it as extensively as I can, and I see it's been in milhouse's builds as well. Almost guarantied some edge case is going to pop up though:\ so keep an eye out!
